File: domain/src/main/java/com/ricamgar/notify/domain/reminder/usecase/GetReminders.java
package com.ricamgar.notify.domain.reminder.usecase;

import com.ricamgar.notify.domain.base.usecase.UseCase;
import com.ricamgar.notify.domain.reminder.model.Reminder;
import com.ricamgar.notify.domain.reminder.repository.RemindersRepository;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import rx.Observable;
import rx.Scheduler;

public class GetReminders extends UseCase<List<Reminder>> {

    private final RemindersRepository remindersRepository;
    private boolean history;

    @Inject
    protected GetReminders(RemindersRepository remindersRepository,
                           @Named(value = "ioThread") Scheduler executionThread,
                           @Named(value = "mainThread") Scheduler postExecutionThread) {
        super(executionThread, postExecutionThread);
        this.remindersRepository = remindersRepository;
    }

    @Override
    protected Observable<List<Reminder>> buildUseCaseObservable() {
        if (history) {
            return remindersRepository.getHistoryReminders();
        } else {
            return remindersRepository.getTodoReminders();
        }
    }

    public GetReminders history(boolean history) {
        this.history = history;
        return this;
    }
}

File: domain/src/main/java/com/ricamgar/notify/domain/reminder/usecase/GetReminders.kt
package com.ricamgar.notify.domain.reminder.usecase

import com.ricamgar.notify.domain.reminder.model.Reminder
import com.ricamgar.notify.domain.reminder.repository.RemindersRepository
import io.reactivex.Flowable
import javax.inject.Inject

class GetReminders
@Inject constructor(private val remindersRepository: RemindersRepository) {

  fun execute(history: Boolean): Flowable<List<Reminder>> = when {
    history -> remindersRepository.historyReminders
    else -> remindersRepository.todoReminders
  }
}
